fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/ReplacePredicate.java

```diff
@@ -0,0 +1,321 @@
+            if (not.child(0) instanceof InPredicate && validInPredicate((InPredicate) not.child(0))
+                    || not.child(0) instanceof ComparisonPredicate
+                    && validComparisonPredicate((ComparisonPredicate) not.child(0))) {
+                for (Expression expr : getAllSubExpressions(not.child(0).child(0))) {
+                    context.computeIfAbsent(expr, k -> new LinkedHashSet<>()).add(not);
+                }
+            }
+            return null;
+        }
+
+        @Override
+        public Void visitLike(Like like, Map<Expression, Set<Expression>> context) {
+            if (!(like.child(1) instanceof Literal)) {
+                return null;
+            }
+            for (Expression expr : getAllSubExpressions(like.child(0))) {
+                context.computeIfAbsent(expr, k -> new LinkedHashSet<>()).add(like);
+            }
+            return null;
+        }
+
+        private boolean validComparisonPredicate(ComparisonPredicate comparisonPredicate) {
+            return comparisonPredicate.right() instanceof Literal;
+        }
+
+        private boolean validInPredicate(InPredicate inPredicate) {
+            return inPredicate.isLiteralChildren();
+        }
+    }
+
+    /* replaceToThis: find all predicates that replaceToThis can deduce (e.g. replaceToThis = b)
+     equalSet: the equivalent set of replaceToThis (e.g. equalSet: a=b)
+     exprPredicates: expression and all its corresponding predicates (e.g. such as {a: [a<10, a>1], b: [b in (1, 2)]})
+     return: all predicates that replaceToThis can deduce (return b<10, b>1) */
+    private static <T extends Expression> Set<Expression> getEqualSetAndDoReplace(T replaceToThis, Set<T> equalSet,
+            Map<? extends Expression, Set<Expression>> exprPredicates) {
+        ExpressionAnalyzer analyzer = new ReplaceAnalyzer(null, new Scope(ImmutableList.of()), null, false, false);
+        Set<Expression> res = new LinkedHashSet<>();
+        for (T equals : equalSet) {
+            Map<Expression, Expression> replaceMap = new HashMap<>();
+            replaceMap.put(equals, replaceToThis);
+            if (!exprPredicates.containsKey(equals)) {
+                continue;
+            }
+            for (Expression predicate : exprPredicates.get(equals)) {
+                Expression newPredicates = Replacer.INSTANCE.replace(predicate, replaceMap);
+                try {
+                    Expression analyzed = analyzer.analyze(newPredicates);
+                    res.add(analyzed.withInferred(true));
+                } catch (Exception e) {
+                    // has cast error, just not infer and do nothing
+                    continue;
+                }
+            }
+        }
+        return res;
+    }
+
+    /* Extract the equivalence relationship a=b, and when case (d_tinyint as int)=d_int is encountered,
+    remove the cast and extract d_tinyint=d_int
```

Comment:
could we just use a set instead of a pair, to record the equality of multiple expr, such as a set with c1, c2, c3, to avoid using multiple entry to record c1, c2, c2?

Reply:
why impl the ImmutableEqualSet using a map? looks a little strange...